File: src/app/commands/date_cmd.py
"""date command - list date formats for selection and paste.

Usage in Alfred:  date
                  date <filter>   — filter by format name or example
"""

from __future__ import annotations

import datetime
from typing import NamedTuple

from alfred.logger import get_logger
from alfred.response import item, output

log = get_logger(__name__)


class DateFormat(NamedTuple):
    uid: str
    label: str
    strftime: str | None
    custom: str | None = None  # used when strftime is not sufficient


def _unix_timestamp(dt: datetime.datetime) -> str:
    return str(int(dt.timestamp()))


_FORMATS: list[DateFormat] = [
    DateFormat("yyyymmdd", "YYYYMMDD", "%Y%m%d"),
    DateFormat("yymmdd", "YYMMDD", "%y%m%d"),
    DateFormat("iso-date", "YYYY-MM-DD", "%Y-%m-%d"),
    DateFormat("slash-ymd", "YYYY/MM/DD", "%Y/%m/%d"),
    DateFormat("slash-mdy", "MM/DD/YYYY", "%m/%d/%Y"),
    DateFormat("slash-dmy", "DD/MM/YYYY", "%d/%m/%Y"),
    DateFormat("abbr-month", "MMM DD, YYYY", "%b %d, %Y"),
    DateFormat("full-month", "MMMM DD, YYYY", "%B %d, %Y"),
    DateFormat("iso-datetime", "YYYY-MM-DDThh:mm:ss", "%Y-%m-%dT%H:%M:%S"),
    DateFormat("unix", "Unix timestamp", None),
]


def _formatted(fmt: DateFormat, dt: datetime.datetime) -> str:
    if fmt.uid == "unix":
        return _unix_timestamp(dt)
    assert fmt.strftime is not None
    return dt.strftime(fmt.strftime)


def handle(args: str) -> None:
    """Show date formats; optionally filter by format label or example value."""
    log.debug("date command: args=%r", args)

    now = datetime.datetime.now()
    query = args.strip().lower()

    items = []
    for fmt in _FORMATS:
        value = _formatted(fmt, now)
        if query and not (query in fmt.label.lower() or query in value.lower() or query in fmt.uid):
            continue
        items.append(
            item(
                title=value,
                subtitle=fmt.label,
                arg=value,
                uid=fmt.uid,
            )
        )

    if not items:
        output(
            [
                item(
                    title=f'No format matches "{args}"',
                    subtitle="Try: YYYY, MM, DD, unix, ISO ...",
                    valid=False,
                )
            ]
        )
        return

    output(items)

File: internal/datecmd/datecmd.go
// Package datecmd dispatches an Alfred "date" query to the matching
// command handler and builds the Script Filter response.
//
// Commands:
//
//	date                  — today in all formats (default)
//	date -2  / -2d        — 2 days ago
//	date +1w              — 1 week later
//	date -3m              — 3 months ago
//	date +1y              — 1 year later
//	date 2026/7/9         — specific date (4- or 2-digit year, / or - separator)
//	date <filter>         — filter by format name or value (e.g. "ISO", "YYYY", "unix")
//	date help             — show available commands
package datecmd

import (
	"fmt"
	"strconv"
	"strings"
	"unicode"

	"github.com/y-marui/alfred-paste-formatted-date/internal/datefmt"
	"github.com/y-marui/alfred-paste-formatted-date/internal/dateresolve"
	"github.com/y-marui/alfred-paste-formatted-date/internal/scriptfilter"
)

// Dispatch parses the raw Alfred query and routes it to the matching
// command, falling back to the date command when the leading token isn't a
// registered command name (e.g. "ISO" or "-2d").
func Dispatch(query string) scriptfilter.Response {
	trimmed := strings.TrimSpace(query)
	command, rest := splitCommand(trimmed)

	switch strings.ToLower(command) {
	case "help":
		return handleHelp()
	case "date":
		return handleDate(rest)
	default:
		return handleDate(trimmed)
	}
}

// splitCommand splits s into its first whitespace-delimited token and the
// remainder, with the remainder's leading whitespace stripped — mirroring
// Python's "s.strip().split(None, 1)".
func splitCommand(s string) (command, rest string) {
	i := strings.IndexFunc(s, unicode.IsSpace)
	if i < 0 {
		return s, ""
	}
	return s[:i], strings.TrimLeftFunc(s[i:], unicode.IsSpace)
}

// handleDate resolves the target date from args, then shows matching date formats.
func handleDate(args string) scriptfilter.Response {
	targetDT, filterQuery := dateresolve.Resolve(args)
	query := strings.ToLower(strings.TrimSpace(filterQuery))

	var items []scriptfilter.Item
	for _, f := range datefmt.All {
		value := datefmt.Value(f, targetDT)
		if query != "" &&
			!strings.Contains(strings.ToLower(f.Label), query) &&
			!strings.Contains(strings.ToLower(value), query) &&
			!strings.Contains(f.UID, query) {
			continue
		}
		items = append(items, scriptfilter.Item{
			Title:    value,
			Subtitle: f.Label,
			Arg:      value,
			UID:      f.UID,
			Valid:    scriptfilter.BoolPtr(true),
		})
	}

	if len(items) == 0 {
		return scriptfilter.Response{
			Items: []scriptfilter.Item{{
				Title:    fmt.Sprintf(`No format matches "%s"`, args),
				Subtitle: "Try: YYYY, MM, DD, unix, ISO, -2d, +1w, 2026/7/9 ...",
				Valid:    scriptfilter.BoolPtr(false),
			}},
		}
	}
	return scriptfilter.Response{Items: items}
}

type helpEntry struct {
	cmd, desc, autocomplete string
}

var helpCommands = []helpEntry{
	{"date", "List all date formats (default command)", "date "},
	{"date help", "Show this help", "date help"},
}

// handleHelp displays all available commands.
func handleHelp() scriptfilter.Response {
	items := make([]scriptfilter.Item, len(helpCommands))
	for i, c := range helpCommands {
		items[i] = scriptfilter.Item{
			Title:        c.cmd,
			Subtitle:     c.desc,
			UID:          "help-" + strconv.Itoa(i),
			Valid:        scriptfilter.BoolPtr(false),
			Autocomplete: c.autocomplete,
		}
	}
	return scriptfilter.Response{Items: items}
}
